src/client/Microsoft.Identity.Client/ManagedIdentity/ManagedIdentityKeyProviderFactory.cs

```diff
@@ -0,0 +1,69 @@
+﻿// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+using System;
+using System.Threading;
+using Microsoft.Identity.Client.ManagedIdentity.KeyProviders;
+using Microsoft.Identity.Client;
+using Microsoft.Identity.Client.PlatformsCommon.Shared;
+using Microsoft.Identity.Client.Core;
+
+namespace Microsoft.Identity.Client.ManagedIdentity
+{
+    /// <summary>
+    /// Creates (once) and caches the most suitable Managed Identity key provider for the current platform.
+    /// Thread-safe, lock-free (uses CompareExchange).
+    /// </summary>
+    internal static class ManagedIdentityKeyProviderFactory
+    {
+        // Cached singleton instance of the chosen key provider.
+        private static IManagedIdentityKeyProvider s_provider;
+
+        /// <summary>
+        /// Returns the cached provider if available; otherwise creates it in a thread-safe manner.
+        /// </summary>
+        internal static IManagedIdentityKeyProvider GetOrCreateProvider(ILoggerAdapter logger)
+        {
+            // Fast path: read the field once (Volatile ensures latest published value).
+            IManagedIdentityKeyProvider existing = Volatile.Read(ref s_provider);
```

Comment:
Why aren't you using semaphoreSlim or declaring a volatile variable like you do in the other files?